pynumaflow/sinker/servicer/sync_servicer.py

```diff
@@ -1,69 +1,104 @@
-from collections.abc import Iterator, Iterable
+from collections.abc import Iterable
 
-from google.protobuf import empty_pb2 as _empty_pb2
-from pynumaflow._constants import _LOGGER
+
+from pynumaflow._constants import _LOGGER, STREAM_EOF
+from pynumaflow.proto.sinker import sink_pb2_grpc, sink_pb2
 from pynumaflow.shared.server import exit_on_error
-from pynumaflow.sinker._dtypes import Datum
+from pynumaflow.shared.servicer import is_valid_handshake
+from pynumaflow.shared.synciter import SyncIterator
+from pynumaflow.shared.thread_with_return import ThreadWithReturnValue
 from pynumaflow.sinker._dtypes import SyncSinkCallable
-from pynumaflow.proto.sinker import sink_pb2_grpc, sink_pb2
-from pynumaflow.sinker.servicer.utils import build_sink_response
+from pynumaflow.sinker.servicer.utils import (
+    datum_from_sink_req,
+    _create_read_handshake_response,
+    build_sink_resp_results,
+)
 from pynumaflow.types import NumaflowServicerContext
 
 
-def datum_generator(request_iterator: Iterable[sink_pb2.SinkRequest]) -> Iterable[Datum]:
-    for d in request_iterator:
-        datum = Datum(
-            keys=list(d.keys),
-            sink_msg_id=d.id,
-            value=d.value,
-            event_time=d.event_time.ToDatetime(),
-            watermark=d.watermark.ToDatetime(),
-            headers=dict(d.headers),
-        )
-        yield datum
-
-
 class SyncSinkServicer(sink_pb2_grpc.SinkServicer):
     """
     This class is used to create a new grpc Sink servicer instance.
     It implements the SinkServicer interface from the proto sink.proto file.
     Provides the functionality for the required rpc methods.
     """
 
-    def __init__(
-        self,
-        handler: SyncSinkCallable,
-    ):
-        self.__sink_handler: SyncSinkCallable = handler
+    def __init__(self, handler: SyncSinkCallable):
+        self.handler: SyncSinkCallable = handler
 
     def SinkFn(
-        self, request_iterator: Iterator[sink_pb2.SinkRequest], context: NumaflowServicerContext
-    ) -> sink_pb2.SinkResponse:
+        self, request_iterator: Iterable[sink_pb2.SinkRequest], context: NumaflowServicerContext
+    ) -> Iterable[sink_pb2.SinkResponse]:
         """
-        Applies a sink function to a list of datum elements.
-        The pascal case function name comes from the proto sink_pb2_grpc.py file.
+        Applies a sink function to datum elements.
         """
-        # if there is an exception, we will mark all the responses as a failure
-        datum_iterator = datum_generator(request_iterator)
+
         try:
-            rspns = self.__sink_handler(datum_iterator)
+            # The first message to be received should be a valid handshake
+            req = next(request_iterator)
```

Comment:
Might be a good idea to handle the `StopIteration` error differently

Reply:
Should be covered by the Base exception right now, but will think about it and update in a follow up.